### DISCUSSION TITLE
Multiple Analog Inputs Having 1 Axis Output

### DISCUSSION THREAD

**Discussion by MikeFindlay** [fasteddy516/CircuitPython_JoystickXL#9] [Q&A] [answered]
I was wondering if there was a way to have 2 analog inputs (ie: a split throttle handle, 1 for majority of the throttle range, 1 for fine tuning) output into 1 virtual axis. 
Handle 1 influences 75% of the range and Handle 2 would do the remaining 25%.
throttle = (t1 * .75) + (t2 * .25)

Current workarounds I've tried have led to the pins not updating and conflicts due to value, integer, and float compatibility.

**Comment by fasteddy516** [comment]
This sounds like an interesting use case; it's definitely *not* something that is built into JoystickXL's standard inputs, but it should be achievable using an axis with a `VirtualInput` source along with a couple of `Axis` objects with analog inputs (or raw analog reads if you want to keep the two component axes strictly internal).

I've got some time over the next day or two, and will try to come up with a working example.  If you're willing to share which CircuitPython board you're developing for (and which analog inputs you're using), I can try to stick to those in the example.

**Reply by MikeFindlay** [reply]
Thanks man! I was using A1 and A0 on a rpi pico. 
Here's the code I ended up with after trying for a day or so. I was also modifying the inputs file to create a function to force update the virtual inputs.

`from analogio import AnalogIn
import board
from joystick_xl.inputs import Axis, Button, Hat, VirtualInput
from joystick_xl.joystick import Joystick
joystick = Joystick()
vinput = VirtualInput(0)

joystick.add_input(
    Axis(vinput),
)

while True:
    power1 = board.A1.value()
    power2 = board.A0.value()
    power = int((power1 * .75) + (power2 * .25))
    vinput.virupdate(power)
    joystick.update()
#    print (power)
print(type(board.A0))`


`class VirtualInput:
    """Provide an object with a .value property to represent a remote input."""

    def __init__(self, value: Union[bool, int]) -> None:
        """
        Provide an object with a ``.value`` property to represent a remote input.

        :param value: Sets the initial ``.value`` property (Should be ``True`` for
            active-low buttons, ``32768`` for idle/centered axes).
        :type value: bool or int
        """
        self.value = value
    def virupdate(self, newval):
        self.value = newval`

**Comment by fasteddy516** [comment]
It looks like you were pretty close, but were trying to do your analog reads right off of a `board.Pin` object rather than an `AnalogIn`.  In CircuitPython, you have to create `AnalogIn` objects first:

```python
power1 = AnalogIn(board.A1)
power2 = AnalogIn(board.A0)
```

Then read the values off of those objects when needed:

```python
power = int((power1.value * 0.75) + (power2.value * 0.25))
```

The `VirtualInput.value` property should also be usable as-is - it's completely unprotected (for better or worse), so you can technically set the value to anything, although anything other than an `int` or `bool` will upset input processing further up the chain.  The preferred method for manually setting inputs with `VirtualInput` sources is to use the `source_value` property of the input itself.  Here's an example I tested on an rpi pico here:

```python
"""This is a simple compound axis example using JoystickXL."""

import board
from analogio import AnalogIn
from joystick_xl.inputs import Axis, VirtualInput
from joystick_xl.joystick import Joystick

p_coarse = AnalogIn(board.A1)  # analog input used for coarse (75%) range adjustment
p_fine = AnalogIn(board.A0)  # analog input used for fine (25%) range adjustment

joystick = Joystick()

joystick.add_input(
    Axis(source=VirtualInput(value=0)),  # x-axis using a virtual input
)

while True:
    # Here we're manually generating the combined axis value using our two analog
    # inputs.  We're accessing our joystick's `Axis` collection directly, and setting
    # the X-axis `source_value` property, which can be done when an input's source is
    # a `VirtualInput` object.

    joystick.axis[Axis.X].source_value = int(
        (p_coarse.value * 0.75) + (p_fine.value * 0.25)
    )

    joystick.update()
```

Hopefully that helps - and feel free to share anything you make using JoystickXL in the Show & Tell section here!

**Comment by fasteddy516** [comment, marked as answer]
Just in case - here's another example if you actually want all of the axes to report over USB-HID.  Using this code, the *coarse* axis reports as *X* at full scale, the *fine* axis reports as *Y* at full scale, and the combined coarse+fine axis reports as *Z*.

```python
"""Another JoystickXL compound axis example where all axes report over USB-HID."""

import board
from joystick_xl.inputs import Axis, VirtualInput
from joystick_xl.joystick import Joystick

joystick = Joystick()

joystick.add_input(
    Axis(source=board.A1),  # x-axis (coarse)
    Axis(source=board.A0),  # y-axis (fine)
    Axis(source=VirtualInput(value=0)),  # z-axis (compound 0.75x + 0.25y)
)

while True:
    joystick.axis[Axis.Z].source_value = int(
        (joystick.axis[Axis.X].source_value * 0.75)
        + (joystick.axis[Axis.Y].source_value * 0.25)
    )

    joystick.update()
```

**Reply by MikeFindlay** [reply]
Amazing! Thank you so much! I was damn close too. I'm used to matlab for python programming, so I've still got quite a bit to learn in regular python. I'm about 90% done designing a throttle for Star Citizen Mining, so when i drop that, I'll make sure to link your library and give ya credit!
